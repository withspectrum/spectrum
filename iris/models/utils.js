const { db } = require('./db');

export const NEW_DOCUMENTS = db
  .row('old_val')
  .eq(null)
  .and(db.not(db.row('new_val').eq(null)));

export const listenToNewDocumentsIn = (table, cb) => {
  return (
    db
      .table(table)
      .changes({
        includeInitial: false,
      })
      // Filter to only include newly inserted messages in the changefeed
      .filter(NEW_DOCUMENTS)
      .run({ cursor: true }, (err, cursor) => {
        if (err) throw err;
        cursor.each((err, data) => {
          if (err) throw err;
          // Call the passed callback with the message directly
          cb(data.new_val);
        });
      })
  );
};

export const parseRange = timeframe => {
  switch (timeframe) {
    case 'daily': {
      return { current: 60 * 60 * 24, previous: 60 * 60 * 24 * 2 };
    }
    case 'weekly': {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
    case 'monthly': {
      return { current: 60 * 60 * 24 * 30, previous: 60 * 60 * 24 * 60 };
    }
    case 'quarterly': {
      return { current: 60 * 60 * 24 * 90, previous: 60 * 60 * 24 * 180 };
    }
    default: {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
  }
};

export const getAu = (range: string) => {
  const { current } = parseRange(range);
  return db
    .table('users')
    .filter(db.row('lastSeen').during(db.now().sub(current), db.now()))
    .count()
    .default(0)
    .run();
};

export const getGrowth = async (
  table: string,
  range: string,
  field: string,
  filter: ?mixed
) => {
  const { current, previous } = parseRange(range);
  const currentPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(current), db.now()))
    .filter(filter ? filter : '')
    .count()
    .run();

  const prevPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(previous), db.now().sub(current)))
    .filter(filter ? filter : '')
    .count()
    .run();

  const rate = (await (currentPeriodCount - prevPeriodCount)) / prevPeriodCount;
  return {
    currentPeriodCount,
    prevPeriodCount,
    growth: Math.round(rate * 100),
  };
};

export const getCount = (table: string, filter: mixed) => {
  if (filter) {
    return db
      .table(table)
      .filter(filter)
      .count()
      .run();
  }

  return db
    .table(table)
    .count()
    .run();
};
