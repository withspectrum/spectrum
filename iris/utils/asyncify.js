// @flow
// Turn a callback-based listener into an async iterator
const Queue = require('then-queue');

const asyncify = (
  listener: (...args: any[]) => any,
  onError?: Error => void
) => {
  let queues = [];
  listener((...args) => queues.map(q => q.push(...args)));

  return () => {
    const queue = new Queue();
    queues.push(queue);

    // I have no idea how to type generators...
    // $FlowFixMe
    async function* AsyncGenerator() {
      try {
        while (true) {
          yield await queue.pop();
        }
      } catch (err) {
        // Remove the error'ing queue from the queues array
        queues = queues.filter(q => q !== queue);
        onError && onError(err);
      }
    }

    return AsyncGenerator();
  };
};

export default asyncify;
