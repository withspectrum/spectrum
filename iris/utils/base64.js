// @flow

/**
 * Encode a string to base64 (using the Node built-in Buffer)
 *
 * Stolen from http://stackoverflow.com/a/38237610/2115623
 */
export const encode = (string: string) =>
  Buffer.from(string).toString('base64');

type Base64String = string;
/**
 * Decode a base64 string (using the Node built-in Buffer)
 *
 * Stolen from http://stackoverflow.com/a/38237610/2115623
 */
export const decode = (string?: Base64String) =>
  (string ? Buffer.from(string, 'base64').toString('ascii') : '');
