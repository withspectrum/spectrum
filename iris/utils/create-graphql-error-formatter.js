// @flow
import Raven from 'raven';
import { IsUserError } from './UserError';
import type { GraphQLError } from 'graphql';

const createGraphQLErrorFormatter = (req?: express$Request) => (
  error: GraphQLError
) => {
  console.log('error', error);
  const isUserError = error.originalError
    ? error.originalError[IsUserError]
    : false;
  let sentryId = 'ID only generated in production';
  if (!isUserError) {
    if (process.env.NODE_ENV === 'production') {
      sentryId = Raven.captureException(
        error,
        req && Raven.parsers.parseRequest(req)
      );
    }
  }
  return {
    message: isUserError ? error.message : `Internal server error: ${sentryId}`,
    // Hide the stack trace in production mode
    stack:
      !process.env.NODE_ENV === 'production' ? error.stack.split('\n') : null,
  };
};

export default createGraphQLErrorFormatter;
