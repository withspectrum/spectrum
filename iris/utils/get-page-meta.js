const { parse } = require('url');
const generateMetaInfo = require('../../shared/generate-meta-info');

// Don't even try if the path is /<value> any of these
// TODO: Longer, more complete blacklist here
const PATH_BLACKLIST = ['robots.txt', 'home', 'messages', 'notifications'];

type Meta = {
  title: string,
  description: string,
  extra: string,
};

export default (
  url: string,
  request: (query: string) => Promise
): Promise<Meta> => {
  const { pathname = '' } = parse(url);
  const [, first, second, third] = pathname.split('/');
  if (third || first.length === 0) return Promise.resolve(generateMetaInfo());
  let promise;

  switch (first) {
    case 'explore': {
      promise = Promise.resolve(generateMetaInfo({ type: 'explore' }));
      break;
    }
    case 'thread': {
      /**
       * Thread
       */
      promise = request(/* GraphQL */ `
        {
          thread(id: "${second}") {
            content {
              title
              body
            }
            type
            channel {
              name
              isPrivate
            }
          }
        }
      `).then(res => {
        const { thread: { type, content, channel } } = res.data;
        return generateMetaInfo({
          type: 'thread',
          data: {
            title: content.title,
            type,
            body: content.body,
            channelName: channel.name,
            privateChannel: channel.isPrivate,
          },
        });
      });
      break;
    }
    case 'users': {
      /**
       * User
       */
      promise = request(/* GraphQL */ `
        {
          user(username: "${second}") {
            name
            username
            description
            # coverPhoto
          }
        }
      `).then(res => {
        const { user } = res.data;
        return generateMetaInfo({
          type: 'user',
          data: {
            username: user.username,
            description: user.description,
            name: user.name,
          },
        });
      });
      break;
    }
    default: {
      const isBlacklisted = PATH_BLACKLIST.includes(first);
      if (second && !isBlacklisted) {
        /**
         * Channel
         */
        promise = request(/* GraphQL */ `
          {
            channel(channelSlug: "${second}", communitySlug: "${first}") {
              name
              description
              isPrivate
              community {
                name
              }
            }
          }
        `).then(res => {
          const { channel, channel: { community } } = res.data;
          return generateMetaInfo({
            type: 'channel',
            data: {
              name: channel.name,
              description: channel.description,
              communityName: community.name,
              private: channel.isPrivate,
            },
          });
        });
      } else if (!isBlacklisted) {
        /**
         * Community
         */
        promise = request(/* GraphQL */ `
          {
            community(slug: "${first}") {
              name
              description
            }
          }
        `).then(res => {
          const { community } = res.data;
          return generateMetaInfo({
            type: 'community',
            data: {
              name: community.name,
              description: community.description,
            },
          });
        });
      }
      break;
    }
  }

  if (!promise) return Promise.resolve(generateMetaInfo());

  return promise.catch(err => {
    console.log(`⚠️ Failed to load metadata for ${url}! ⚠️`);
    console.log(err);
    return generateMetaInfo();
  });
};
