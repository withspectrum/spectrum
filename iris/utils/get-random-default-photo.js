// @flow
const faker = require('faker');

// Helper function to get a random profile and cover photo
const PALETTE = [
  'blue',
  'blush',
  'cool',
  'green',
  'orange',
  'peach',
  'pink',
  'red',
  'teal',
  'violet',
];
export default () => {
  const color = faker.random.arrayElement(PALETTE);
  return {
    profilePhoto: `https://s3.amazonaws.com/spectrum-chat/default_images/profile-${color}.png`,
    coverPhoto: `https://s3.amazonaws.com/spectrum-chat/default_images/cover-${color}.png`,
  };
};
