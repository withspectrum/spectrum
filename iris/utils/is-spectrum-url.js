// @flow
import { URL } from 'url';
const IS_PROD = process.env.NODE_ENV === 'production';
/**
 * Make a URL string is a spectrum.chat URL
 */
export default (url: string): boolean => {
  try {
    const { hostname } = new URL(url);
    // hostname might be spectrum.chat or subdomain.spectrum.chat, so we use .endsWith
    // We don't just check .contains because otherwise folks could make spectrum.chat.mydomain.com
    const IS_SPECTRUM_URL = hostname.endsWith('spectrum.chat');
    const IS_LOCALHOST = hostname === 'localhost';
    // Make sure the passed redirect URL is a spectrum.chat one or (in development) localhost
    if (IS_SPECTRUM_URL || (!IS_PROD && IS_LOCALHOST)) {
      return true;
    }
  } catch (err) {
    // Swallow URL parsing errors (when an invalid URL is passed) and redirect to the standard one
    console.log(`Invalid URL ("${url}") passed. Full error:`);
    console.log(err);
  }
  return false;
};
