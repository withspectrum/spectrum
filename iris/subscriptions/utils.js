// @flow
import { getUserPermissionsInChannel } from '../models/usersChannels';
import { getMembersInDirectMessageThread } from '../models/usersDirectMessageThreads';
import { getChannels } from '../models/channel';
import type { DBThread } from '../models/thread';
import type { GraphQLContext } from '../';

// Check that a user can view a channel
export const userCanViewChannel = async (channelId: string, userId: string) => {
  const [channel] = await getChannels([channelId]);
  // Everybody can view public channels
  if (!channel.isPrivate) return true;

  if (!userId) return false;
  return await userIsMemberOfChannel(channelId, userId);
};

export const userIsMemberOfChannel = async (
  channelId: string,
  userId: string
) => {
  if (!userId) return false;
  const { isMember } = await getUserPermissionsInChannel(channelId, userId);
  return isMember;
};

export const userCanViewDirectMessageThread = async (
  threadId: string,
  userId: string
) => {
  const members = await getMembersInDirectMessageThread(threadId);
  if (!members) return false;
  return members.some(member => member.userId === userId);
};
