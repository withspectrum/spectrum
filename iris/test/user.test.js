import { request } from './utils';

describe('queries', () => {
  it('should fetch a user', () => {
    const query = /* GraphQL */ `
			{
				user(id: "gVk5mYwccUOEKiN5vtOouqroGKo1") {
          id
          name
          description
          website
          username
          email
          providerId
          createdAt
				}
			}
		`;

    expect.assertions(1);
    return request(query).then(result => {
      expect(result).toMatchSnapshot();
    });
  });

  it('should return null for a non-existant id', () => {
    const query = /* GraphQL */ `
			{
				user(id: "non-existant") {
					id
          username
				}
			}
		`;

    expect.assertions(1);
    return request(query).then(result => {
      expect(result).toMatchSnapshot();
    });
  });

  it('should fetch a users communities', () => {
    const query = /* GraphQL */ `
      {
        user(id: "gVk5mYwccUOEKiN5vtOouqroGKo1") {
          communityConnection {
            edges {
              node {
                name
              }
            }
          }
        }
      }
    `;

    expect.assertions(1);
    return request(query).then(result => {
      expect(result).toMatchSnapshot();
    });
  });

  describe.skip('everything', () => {
    it('should return the latest thread', () => {
      const query = /* GraphQL */ `
  			{
  				user(id: "gVk5mYwccUOEKiN5vtOouqroGKo1") {
  					everything(first: 1) {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
  				}
  			}
  		`;
      expect.assertions(1);
      return graphql(schema, query).then(result => {
        expect(result).toMatchSnapshot();
      });
    });

    it('should paginate based on the after property', () => {
      const query = /* GraphQL */ `
  			{
  				user(id: "gVk5mYwccUOEKiN5vtOouqroGKo1") {
  					everything(first: 1, after: "c2Vjb25kLXN0b3J5") {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
  				}
  			}
  		`;
      expect.assertions(1);
      return graphql(schema, query).then(result => {
        expect(result).toMatchSnapshot();
      });
    });

    it('should handle first being set to 0 correctly', () => {
      const noCursorQuery = /* GraphQL */ `
  			{
  				user(id: "gVk5mYwccUOEKiN5vtOouqroGKo1") {
  					everything(first: 0) {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
  				}
  			}
  		`;
      const cursorQuery = /* GraphQL */ `
  			{
  				user(id: "gVk5mYwccUOEKiN5vtOouqroGKo1") {
  					everything(first: 0, after: "Zmlyc3Qtc3Rvcnk=") {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
  				}
  			}
  		`;
      expect.assertions(2);
      return graphql(schema, noCursorQuery)
        .then(result => {
          expect(result).toMatchSnapshot();
          return graphql(schema, cursorQuery);
        })
        .then(result => {
          expect(result).toMatchSnapshot();
        });
    });
  });
});
