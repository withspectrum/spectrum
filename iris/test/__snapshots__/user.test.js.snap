// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries should fetch a user 1`] = `
Object {
  "data": Object {
    "user": Object {
      "coverPhoto": "https://spectrum-imgp.imgix.net/https%3A%2F%2Fpbs.twimg.com%2Fprofile_banners%2F2451223458%2F1479507323%2F1500x500?w=640&h=192&ixlib=js-1.1.0&s=b860478fd198543b582046f1a55d0fab",
      "createdAt": "2016-12-31T23:00:00.000Z",
      "description": "Makes styled-components, react-boilerplate and micro-analytics ðŸ’… Speciality coffee geek, skier, traveller â˜•",
      "email": "contact@mxstbr.com",
      "id": "gVk5mYwccUOEKiN5vtOouqroGKo1",
      "lastSeen": "2016-12-31T23:00:00.000Z",
      "name": "Max Stoiber",
      "profilePhoto": "https://spectrum-imgp.imgix.net/https%3A%2F%2Fimg.gs%2Fjztmrqvgzv%2F500%2Fmxstbr.com%2Fheadshot.jpeg?w=128&h=128&ixlib=js-1.1.0&s=9b6e886953006a3441316d73683e7168",
      "providerId": "2451223458",
      "username": "mxstbr",
      "website": "https://mxstbr.com",
    },
  },
}
`;

exports[`queries should return null for a non-existant id 1`] = `
Object {
  "data": Object {
    "user": null,
  },
}
`;
